clients/src/main/java/org/apache/kafka/common/Cluster.java

```diff
@@ -108,23 +108,57 @@ private Cluster(String clusterId,
 
         // Index the nodes for quick lookup
         Map<Integer, Node> tmpNodesById = new HashMap<>();
-        for (Node node : nodes)
+        Map<Integer, List<PartitionInfo>> tmpPartitionsByNode = new HashMap<>(nodes.size());
```

Comment:
`tmpPartitionsByNode` has modifiable lists as values and by extension `this.partitionsByNode` will as well. Is that a concern? We can access it via a public interface to be fair

Reply:
Yeah, it's a bit annoying that we have to do an iteration just to do this. Pushed a fix.

Reply:
I also added a test to verify this since the logic is a bit complex.

Reply:
nice